*AIRBNB PROJECT*
*BY NHI BUI*

## 1. Load the data
Loading the Airbnb NYC Open Data from Kaggle with around 102K listings and 26 columns.

In [17]:
pip -q install ipykernel pandas numpy matplotlib seaborn statsmodels scipy


Note: you may need to restart the kernel to use updated packages.


In [18]:
import pandas as pd

## 2. Understand the data
Import data from uploaded Airbnb_Open_Data.csv abd read the data overview to check its quality and format

In [71]:
df = pd.read_csv('../data/Airbnb_Open_Data.csv', low_memory=False)
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 102599 entries, 0 to 102598
Data columns (total 26 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   id                              102599 non-null  int64  
 1   NAME                            102349 non-null  str    
 2   host id                         102599 non-null  int64  
 3   host_identity_verified          102310 non-null  str    
 4   host name                       102193 non-null  str    
 5   neighbourhood group             102570 non-null  str    
 6   neighbourhood                   102583 non-null  str    
 7   lat                             102591 non-null  float64
 8   long                            102591 non-null  float64
 9   country                         102067 non-null  str    
 10  country code                    102468 non-null  str    
 11  instant_bookable                102494 non-null  object 
 12  cancellation_policy        

In [72]:
df.describe()

,id,host id,lat,long,Construction year,minimum nights,number of reviews,reviews per month,review rate number,calculated host listings count,availability 365
count,1.025990e+05,1.025990e+05,102591.000000,102591.000000,102385.000000,102190.000000,102416.000000,86720.000000,102273.000000,102280.000000,102151.000000
mean,2.914623e+07,4.925411e+10,40.728094,-73.949644,2012.487464,8.135845,27.483743,1.374022,3.279106,7.936605,141.133254
std,1.625751e+07,2.853900e+10,0.055857,0.049521,5.765556,30.553781,49.508954,1.746621,1.284657,32.218780,135.435024
min,1.001254e+06,1.236005e+08,40.499790,-74.249840,2003.000000,-1223.000000,0.000000,0.010000,1.000000,1.000000,-10.000000
25%,1.508581e+07,2.458333e+10,40.688740,-73.982580,2007.000000,2.000000,1.000000,0.220000,2.000000,1.000000,3.000000
50%,2.913660e+07,4.911774e+10,40.722290,-73.954440,2012.000000,3.000000,7.000000,0.740000,3.000000,1.000000,96.000000
75%,4.320120e+07,7.399650e+10,40.762760,-73.932350,2017.000000,5.000000,30.000000,2.000000,4.000000,2.000000,269.000000
max,5.736742e+07,9.876313e+10,40.916970,-73.705220,2022.000000,5645.000000,1024.000000,90.000000,5.000000,332.000000,3677.000000


In [73]:
#Find top 5 AirBnB
df.head()

,id,NAME,host id,host_identity_verified,host name,neighbourhood group,neighbourhood,lat,long,country,...,service fee,minimum nights,number of reviews,last review,reviews per month,review rate number,calculated host listings count,availability 365,house_rules,license
0,1001254,Clean & quiet apt home by the park,80014485718,unconfirmed,Madaline,Brooklyn,Kensington,40.64749,-73.97237,United States,...,$193,10.0,9.0,10/19/2021,0.21,4.0,6.0,286.0,Clean up and treat the home the way you'd like...,NaN
1,1002102,Skylit Midtown Castle,52335172823,verified,Jenna,Manhattan,Midtown,40.75362,-73.98377,United States,...,$28,30.0,45.0,5/21/2022,0.38,4.0,2.0,228.0,Pet friendly but please confirm with me if the...,NaN
2,1002403,THE VILLAGE OF HARLEM....NEW YORK !,78829239556,NaN,Elise,Manhattan,Harlem,40.80902,-73.94190,United States,...,$124,3.0,0.0,NaN,NaN,5.0,1.0,352.0,"I encourage you to use my kitchen, cooking and...",NaN
3,1002755,NaN,85098326012,unconfirmed,Garry,Brooklyn,Clinton Hill,40.68514,-73.95976,United States,...,$74,30.0,270.0,7/5/2019,4.64,4.0,1.0,322.0,NaN,NaN
4,1003689,Entire Apt: Spacious Studio/Loft by central park,92037596077,verified,Lyndon,Manhattan,East Harlem,40.79851,-73.94399,United States,...,$41,10.0,9.0,11/19/2018,0.10,3.0,1.0,289.0,"Please no smoking in the house, porch or on th...",NaN


## 3. Clean data
Check for missing values and count total missing values from each column in descending order

In [74]:
df.isnull().sum().sort_values(ascending=False)

license                           102597
house_rules                        52131
last review                        15893
reviews per month                  15879
country                              532
availability 365                     448
minimum nights                       409
host name                            406
review rate number                   326
calculated host listings count       319
host_identity_verified               289
service fee                          273
NAME                                 250
price                                247
Construction year                    214
number of reviews                    183
country code                         131
instant_bookable                     105
cancellation_policy                   76
neighbourhood group                   29
neighbourhood                         16
long                                   8
lat                                    8
id                                     0
host id         

## Comment
* license → 102,597 nulls out of 102,599. Basically empty => redundant
* house_rules → half missing => not really needed 
* country / country code will be the same since it's all NYC => not needed column
* 102599 entries and 26 columns
* price and service fees should be numbers instead of strings
* Because some rows have number of reviews equals 0 => its last review should also be NaN and the reviews per month should also show as 0 instead of NaN

In [75]:
#1. Drop redundant columns
df.drop(columns=['country','country code', 'house_rules', 'license', 'host name'], inplace = True)

In [76]:
#2. Drop rows with fewer than 50 missing values 
df.dropna(subset = ['lat', 'long', 'neighbourhood', 'neighbourhood group', 'cancellation_policy', 'instant_bookable', 'number of reviews', 'Construction year', 'price'], inplace = True)

In [77]:
#3. Change data type of service fee and price
df['service fee']=df['service fee'].replace('[\$,]', "", regex=True).astype(float)
df['price']=df['price'].replace(r'[\$,]', "", regex=True).astype(float)

<>:2: SyntaxWarning: "\$" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\$"? A raw string is also an option.
<>:2: SyntaxWarning: "\$" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\$"? A raw string is also an option.
C:\Users\Admin\AppData\Local\Temp\ipykernel_28948\1773299035.py:2: SyntaxWarning: "\$" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\$"? A raw string is also an option.
  df['service fee']=df['service fee'].replace('[\$,]', "", regex=True).astype(float)


In [78]:
#4. Change the reviews per month to 0 for listings with no reviews
df['reviews per month'] = df['reviews per month'].fillna(0)

In [79]:
df.isnull().sum().sort_values(ascending=False)


last review                       15698
minimum nights                      386
availability 365                    359
calculated host listings count      316
review rate number                  301
host_identity_verified              258
NAME                                232
service fee                         229
id                                    0
neighbourhood group                   0
host id                               0
neighbourhood                         0
lat                                   0
price                                 0
Construction year                     0
room type                             0
cancellation_policy                   0
instant_bookable                      0
long                                  0
number of reviews                     0
reviews per month                     0
dtype: int64

* looking at the rest of the columns of missing values, now we need to analyze if these columns are important to our analysis or now 

In [80]:
df[df['last review'].isnull()].describe()

,id,host id,lat,long,Construction year,price,service fee,minimum nights,number of reviews,reviews per month,review rate number,calculated host listings count,availability 365
count,1.569800e+04,1.569800e+04,15698.000000,15698.000000,15698.000000,15698.000000,15676.000000,15638.000000,15698.000000,15698.000000,15663.000000,15640.000000,15492.000000
mean,2.756264e+07,4.954346e+10,40.731749,-73.955717,2012.494012,621.780227,124.372225,11.777977,0.110906,0.000657,3.279576,12.991496,135.873548
std,1.609618e+07,2.858970e+10,0.053040,0.044518,5.771553,331.746881,66.329268,41.988110,4.130538,0.031299,1.291820,44.591357,143.274974
min,1.002403e+06,1.305934e+08,40.499790,-74.242850,2003.000000,50.000000,10.000000,-1223.000000,0.000000,0.000000,1.000000,1.000000,-10.000000
25%,1.397155e+07,2.505937e+10,40.696448,-73.984470,2008.000000,336.000000,67.000000,1.000000,0.000000,0.000000,2.000000,1.000000,0.000000
50%,2.581424e+07,4.944745e+10,40.728115,-73.959740,2012.000000,621.000000,124.000000,3.000000,0.000000,0.000000,3.000000,1.000000,83.500000
75%,4.331815e+07,7.459097e+10,40.763512,-73.939765,2018.000000,910.000000,182.000000,12.000000,0.000000,0.000000,4.000000,2.000000,282.000000
max,5.735637e+07,9.876097e+10,40.911690,-73.716900,2022.000000,1200.000000,240.000000,3455.000000,222.000000,2.120000,5.000000,327.000000,426.000000


In [83]:
df['service fee'].describe()

count    101597.000000
mean        125.055130
std          66.340524
min          10.000000
25%          68.000000
50%         125.000000
75%         183.000000
max         240.000000
Name: service fee, dtype: float64

In [84]:
df['price'].describe()

count    101826.000000
mean        625.368884
std         331.708664
min          50.000000
25%         340.000000
50%         624.000000
75%         913.000000
max        1200.000000
Name: price, dtype: float64

In [85]:
df['minimum nights'].describe()

count    101440.000000
mean          8.127938
std          30.595460
min       -1223.000000
25%           2.000000
50%           3.000000
75%           5.000000
max        5645.000000
Name: minimum nights, dtype: float64

The listings with no last review (about 15% of the total dataset) has a higher min number of nights and fewer availability dates but around the same service fee and price
=> these hosts want to maximize earnings during peak season or major events but lower avaialbility dates
=> unrealistic expectations during off season time 
=> decide to drop these listings because it may distract from the real data pattern and also these may just be churned listing that never got any bookings

In [86]:
#5. Drop rows with missing values in the last review column
df.dropna(subset = 'last review', inplace = True)

for the rest of columns, since i dont need to analyze data about host, i dropped host name, calculated host listings count and host identity verified 

In [87]:
#6. Drop redundant columns related to host
df.drop(columns = ['host_identity_verified', 'calculated host listings count'], inplace = True)

In [89]:
df.isnull().sum().sort_values(ascending=False)

minimum nights         326
review rate number     266
service fee            207
NAME                   197
availability 365       153
host id                  0
id                       0
neighbourhood group      0
neighbourhood            0
lat                      0
long                     0
Construction year        0
room type                0
cancellation_policy      0
instant_bookable         0
number of reviews        0
price                    0
reviews per month        0
last review              0
dtype: int64

In [91]:
df.fillna(0, inplace = True)

TypeError: Invalid value '0' for dtype 'str'. Value should be a string or missing value, got 'int' instead.